packages/sources/oanda/src/transport/modifiedSSE.ts

```diff
@@ -0,0 +1,178 @@
+import { EndpointContext } from '@chainlink/external-adapter-framework/adapter'
+import { AdapterConfig } from '@chainlink/external-adapter-framework/config'
+import {
+  makeLogger,
+  sleep,
+  PartialSuccessfulResponse,
+  ProviderResult,
+  TimestampedProviderResult,
+} from '@chainlink/external-adapter-framework/util'
+import {
+  TransportDependencies,
+  TransportGenerics,
+} from '@chainlink/external-adapter-framework/transports'
+import {
+  StreamingTransport,
+  SubscriptionDeltas,
+} from '@chainlink/external-adapter-framework/transports/abstract/streaming'
+
+import axios from 'axios'
+
+const logger = makeLogger('ModifiedSSETransport')
+
+export type ModifiedSSEConfig = {
+  url: string
+  config?: {
+    responseType: 'stream'
+    headers: { [header: string]: string }
+  }
+}
+
+export class ModifiedSseTransport<T extends TransportGenerics> extends StreamingTransport<T> {
+  eventListeners!: {
+    type: string
+    parseResponse: (eventt: MessageEvent<any>['data']) => ProviderResult<T>[]
+  }[]
+
+  constructor(
+    private config: {
+      prepareSSEConnectionConfig: (
+        subscriptions: SubscriptionDeltas<T['Request']['Params']>['desired'],
+        context: EndpointContext<T>,
+      ) => ModifiedSSEConfig
+      eventListeners: {
+        type: string
+        parseResponse: (event: MessageEvent<any>['data']) => ProviderResult<T>[]
+      }[]
+      parseSubscriptionList: (
+        subscriptions: SubscriptionDeltas<T['Request']['Params']>['desired'],
+        context: EndpointContext<T>,
+      ) => Promise<(string | undefined)[]>
+    },
+  ) {
+    super()
+  }
+
+  getSubscriptionTtlFromConfig(config: AdapterConfig<T['CustomSettings']>): number {
+    return config.SSE_SUBSCRIPTION_TTL
+  }
+
+  override async initialize(
+    dependencies: TransportDependencies<T>,
+    config: AdapterConfig<T['CustomSettings']>,
+    endpointName: string,
+  ): Promise<void> {
+    super.initialize(dependencies, config, endpointName)
+  }
+
+  async streamHandler(
+    context: EndpointContext<T>,
+    subscriptions: SubscriptionDeltas<T['Request']['Params']>,
+  ): Promise<void> {
+    if (subscriptions.new.length || subscriptions.stale.length) {
+      logger.info({ msg: 'Updating SSE subscriptions', subscriptions })
+
+      const subsList = await this.config.parseSubscriptionList(subscriptions.desired, context)
+
+      logger.info({ subsList })
+
+      if (subsList.some((s) => !s)) {
+        logger.error({ msg: `Cannot update SSE subscriptions, subscriptions contains invalid sub` })
+      } else {
+        const sseConfig = this.config.prepareSSEConnectionConfig(subsList as string[], context)
+
+        const providerDataStreamEstablishedUnixMs = Date.now()
+
+        const restartStream = async () => {
```

Comment:
Is the logic to restart the stream on every call to the streamHandler the expected subscription model for Oanda or is this something you just had to do since the default SSE transport wasn't playing nice?

Reply:
As opposed to subscribe/unsubscribe and keep alive calls after establishing the connection once

Reply:
If there are new or stale subscriptions then yes, we'd have to restart the stream. The stream is established with the list of pairs to subscribe to as input params that can't be updated while the stream is active (see `GET /v3/accounts/{accountID}/pricing/stream` [here](https://developer.oanda.com/rest-live-v20/pricing-ep/)).

Reply:
I see. All the more reason you needed a custom transport. Good to know